File: 03. Stack and Queue/03. Next Greater Element Circular/main.js
function nextGreaterElementCircular(nums) {
    let n = nums.length;
    let result = new Array(n).fill(-1);
    let stack = [];

    for (let i = 0; i < 2 * n; i++) {
        while (stack.length > 0 && nums[stack[stack.length - 1]] < nums[i % n]) {
            result[stack.pop()] = nums[i % n];
        }
        if (i < n) stack.push(i);
    }

    return result;
}

// Main function
(function main() {
    let nums = [4, 5, 2, 10, 8];
    let result = nextGreaterElementCircular(nums);

    console.log("Next Greater Elements (Circular):");
    console.log(result);
})();

File: 03. Stack and Queue/03. Next Greater Element Circular/main.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> nextGreaterElementCircular(vector<int>& nums) {
    int n = nums.size();
    vector<int> result(n, -1);
    stack<int> st;

    for (int i = 0; i < 2 * n; i++) {
        while (!st.empty() && nums[st.top()] < nums[i % n]) {
            result[st.top()] = nums[i % n];
            st.pop();
        }
        if (i < n) st.push(i);
    }

    return result;
}

int main() {
    vector<int> nums = {4, 5, 2, 10, 8};
    vector<int> result = nextGreaterElementCircular(nums);

    cout << "Next Greater Elements (Circular):\n";
    for (int num : result) {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}
